File: testing/tests/python/e2e_test.py
#!/usr/bin/env python3

from selenium import webdriver
from selenium.webdriver.common.by import By


class TestLuckyApp:
    """E2E integration tests class."""

    def setup_method(self, method):
        options = webdriver.FirefoxOptions()
        self.driver = webdriver.Remote('http://firefoxdriver:4444/wd/hub', options=options)
        self.driver.set_window_size(1936, 1056)
        self.driver.get("http://proxy")

    def teardown_method(self, method):
        self.driver.quit()

    def test_frontend_app(self):
        """Test landing page of fronted app."""
        element = self.driver.find_element(By.ID, 'recommendation')
        assert element.text == "Pink Flody - Pulse"

        self.driver.save_screenshot("/app/tmp/pyshot1.png")

    def test_movie_app(self):
        """Test backend app(s)."""
        # Click recomendation button.
        element = self.driver.find_element(By.ID, 'btn-recommend-movie')
        element.click()

        # Check if new recomendation appeard on the webpage.
        element = self.driver.find_element(By.ID, 'recommendation')
        assert element.text != "Pink Flody - Pulse"

        self.driver.save_screenshot("/app/tmp/pyshot2.png")

    def test_music_app(self):
        """Test backend app(s)."""
        # Click recomendation button.
        element = self.driver.find_element(By.ID, 'btn-recommend-music')
        element.click()

        # Check if new recomendation appeard on the webpage.
        element = self.driver.find_element(By.ID, 'recommendation')
        assert element.text != "Pink Flody - Pulse"

        self.driver.save_screenshot("/app/tmp/pyshot3.png")

File: testing/tests/javascript/e2e.test.js
// https://www.freecodecamp.org/news/how-to-start-unit-testing-javascript/
const {Builder, By, Capabilities} = require('selenium-webdriver');
const fs = require('fs');

// Create driver
const capabilities = Capabilities.firefox();
const driver = new Builder()
  .forBrowser('firefox')
  .usingServer("http://firefoxdriver:4444/wd/hub")
  .withCapabilities(capabilities)
  .build();

// Set Up
beforeAll(async () => {
  await driver.get("http://proxy");
});

// Tear Down
afterAll(async () => {
  await driver.quit();
});

const printScreen = async (path) =>{
  let encodedString = await driver.takeScreenshot();
  await fs.writeFileSync(path, encodedString, 'base64');
};

test("Tes frontend app", async () => {
  try {
    const text = await driver.findElement(By.id('recommendation')).getText();
    expect(text).toContain("Pink Flody - Pulse");
  } catch (err) {
    throw err;
  } finally {
    await printScreen('/app/tmp/jshot1.png');
  }
});

test("Tes movie app", async () => {
  try {
    // Click recommendation button
    const element = await driver.findElement(By.id('btn-recommend-movie'));
    element.click();

    // Check new recommendation
    const text = await driver.findElement(By.id('recommendation')).getText();
    expect(text).not.toContain("Pink Flody - Pulse");
  } catch (err) {
    throw err;
  } finally {
    await printScreen('/app/tmp/jshot2.png');
  }
});

test("Tes music app", async () => {
  try {
    // Click recommendation button
    const element = await driver.findElement(By.id('btn-recommend-music'));
    element.click();

    // Check new recommendation
    const text = await driver.findElement(By.id('recommendation')).getText();
    expect(text).not.toContain("Pink Flody - Pulse");
  } catch (err) {
    throw err;
  } finally {
    await printScreen('/app/tmp/jshot3.png');
  }
});